x-pack/plugins/observability_solution/apm/server/utils/es_fields_mappings.ts

```diff
@@ -0,0 +1,814 @@
+    },
+    error: {
+      id: normalizeValue<string>(fields[ERROR_ID]),
+      exception: [
+        {
+          message: normalizeValue<string>(fields[ERROR_EXC_MESSAGE]),
+          type: normalizeValue<string>(fields[ERROR_EXC_TYPE]),
+        },
+      ],
+      grouping_key: normalizeValue<string>(fields[ERROR_GROUP_ID]),
+    },
+    processor: {
+      name: normalizeValue<'error'>(fields[PROCESSOR_NAME]),
+      event: normalizeValue<'error'>(fields[PROCESSOR_EVENT]),
+    },
+    transaction: {
+      id: normalizeValue<string>(fields[TRANSACTION_ID]),
+      type: normalizeValue<string>(fields[TRANSACTION_TYPE]),
+      sampled: normalizeValue<boolean>(fields[TRANSACTION_SAMPLED]),
+    },
+    timestamp: {
+      us: normalizeValue<number>(fields[TIMESTAMP]),
+    },
+  };
+};
+
+export const serviceMetadataDetailsMapping = (
+  fields: Partial<Record<string, unknown[]>>
+): ServiceMetadataDetailsRaw | undefined => {
+  if (!fields) return undefined;
+  const kubernetesNamespace = normalizeValue<string>(fields[KUBERNETES_NAMESPACE]);
+  const containerId = normalizeValue<string>(fields[CONTAINER_ID]);
+  const cloudServiceName = normalizeValue<string>(fields[CLOUD_PROVIDER]);
+  const serviceRuntimeName = normalizeValue<string>(fields[SERVICE_RUNTIME_NAME]);
+  return {
+    service: {
+      name: normalizeValue<string>(fields[SERVICE_NAME]),
+      version: normalizeValue<string>(fields[SERVICE_VERSION]),
+      environment: normalizeValue<string>(fields[SERVICE_ENVIRONMENT]),
+      framework: {
+        name: normalizeValue<string>(fields[SERVICE_FRAMEWORK_NAME]),
+        version: normalizeValue<string>(fields[SERVICE_FRAMEWORK_VERSION]),
+      },
+      node: {
+        name: normalizeValue<string>(fields[SERVICE_NODE_NAME]),
+      },
+      ...(serviceRuntimeName
+        ? {
+            runtime: {
+              name: serviceRuntimeName,
+              version: normalizeValue<string>(fields[SERVICE_RUNTIME_VERSION]),
+            },
+          }
+        : undefined),
+      language: {
+        name: normalizeValue<string>(fields[SERVICE_LANGUAGE_NAME]),
+        version: normalizeValue<string>(fields[SERVICE_LANGUAGE_VERSION]),
+      },
+    },
+    agent: {
+      name: normalizeValue<AgentName>(fields[AGENT_NAME]),
+      version: normalizeValue<AgentName>(fields[AGENT_VERSION]),
+    },
+    host: {
+      architecture: normalizeValue<string>(fields[HOST_ARCHITECTURE]),
+      hostname: normalizeValue<string>(fields[HOST_HOSTNAME]),
+      name: normalizeValue<string>(fields[HOST_NAME]),
+      ip: normalizeValue<string>(fields[HOST_IP]),
+      os: {
+        platform: normalizeValue<string>(fields[HOST_OS_PLATFORM]),
+      },
+    },
+    ...(containerId
+      ? {
+          container: {
+            id: containerId,
+            image: normalizeValue<string>(fields[CONTAINER_IMAGE]),
+          },
+        }
+      : undefined),
+    ...(kubernetesNamespace
```

Comment:
not sure if kubernetesNamespace should be the only field to check for here.

What if kubernetesNames does not exist but another Kubernetes field exists, then we wouldn't set it, right?

Maybe we can create a helper function that checks whether any of the nested fields exist. And use that as the precondition for that object to be created instead of relying on a single field. 

The same applies for the other places here as well, such as container, cloud, serviceRuntime ...

The above mentioned helper function could be generic and be reused in all those places.

Reply:
> Maybe we can create a helper function that checks whether any of the nested fields exist.

Maybe something like `anyMatch(fields, "kubernetes.*")`